plugins/optimization/src/main/java/org/qbicc/plugin/opt/ea/EscapeAnalysisIntraMethodBuilder.java

```diff
@@ -0,0 +1,174 @@
+
+public final class EscapeAnalysisIntraMethodBuilder extends DelegatingBasicBlockBuilder  {
+    private final EscapeAnalysisState escapeAnalysisState;
+    private final ConnectionGraph connectionGraph;
+    private final ClassContext bootstrapClassContext;
+
+    public EscapeAnalysisIntraMethodBuilder(final CompilationContext ctxt, final BasicBlockBuilder delegate) {
+        super(delegate);
+        this.connectionGraph = new ConnectionGraph(getCurrentElement().toString());
+        this.escapeAnalysisState = EscapeAnalysisState.get(ctxt);
+        this.escapeAnalysisState.trackMethod(getCurrentElement(), this.connectionGraph);
+        this.bootstrapClassContext = ctxt.getBootstrapClassContext();
+    }
+
+    @Override
+    public Value new_(final ClassObjectType type, final Value typeId, final Value size, final Value align) {
+        final New result = (New) super.new_(type, typeId, size, align);
+
+        connectionGraph.trackNew(result, defaultEscapeValue(type));
+
+        return result;
+    }
+
+    private EscapeValue defaultEscapeValue(ClassObjectType type) {
+        if (isSubtypeOfClass("java/lang/Thread", type) ||
+            isSubtypeOfClass("java/lang/ThreadGroup", type)) {
+            return EscapeValue.GLOBAL_ESCAPE;
+        }
+
+        return EscapeValue.NO_ESCAPE;
+    }
+
+    private boolean isSubtypeOfClass(String name, ClassObjectType type) {
+        return type.isSubtypeOf(bootstrapClassContext.findDefinedType(name).load().getType());
+    }
+
+    @Override
+    public ValueHandle instanceFieldOf(ValueHandle handle, FieldElement field) {
+        final InstanceFieldOf result = (InstanceFieldOf) super.instanceFieldOf(handle, field);
+
+        // T a = new T(...);
+        // To get represent the GC from 'a' to 'new T(...)',
+        // we hijack future 'a' references to fix the pointer.
+        // When 'a.x' is accessed, we fix the pointer from 'a' to 'new T(...)'.
+        handleInstanceFieldOf(result, handle, handle);
+
+        return result;
+    }
+
+    @Override
+    public Node store(ValueHandle handle, Value value, WriteAccessMode mode) {
+        final Node result = super.store(handle, value, mode);
+
+        if (handle instanceof StaticField) {
+            // static T a = ...
+            if (value instanceof NotNull nn) {
+                connectionGraph.trackStoreStaticField(handle, nn.getInput());
+            } else {
+                connectionGraph.trackStoreStaticField(handle, value);
+            }
+        } else if (handle instanceof InstanceFieldOf && value instanceof New) {
+            // p.f = new T(); // where p is a parameter
+            connectionGraph.fixEdgesNew(handle, (New) value);
+        } else if (handle instanceof LocalVariable && value instanceof New) {
+            connectionGraph.trackLocalNew((LocalVariable) handle, (New) value);
+        }
+
+        return result;
+    }
+
+    @Override
+    public Value call(ValueHandle target, List<Value> arguments) {
```

Comment:
You also will need to track all of the other call types too I think.

Reply:
You mean `callNoSideEffects` and `callNoReturn`? Any other mode?

Reply:
And `TailCall`, and also `Invoke`, `TailInvoke`, and `InvokeNoReturn`.